### PR TITLE
Feature/add student to circle

### PR DESCRIPTION
**Pull request by ElizavetaPetrova**

### CODE REVIEW

**Comment by IlyaBat9** [comment]
SCHOOL-411

File: open_schools_platform/organization_management/circles/views.py [hunk 1]
```diff
@@ -108,3 +113,34 @@ def get(self, request, pk):
...
+        invite_serializer = CircleStudentInviteSerializer(data=request.data)
+        invite_serializer.is_valid(raise_exception=True)
+
+        circle = get_circle(filters={"id": pk})
+        parent_phone = invite_serializer.validated_data["parent_phone"]
+        student_phone = invite_serializer.validated_data["student_phone"]
+        email = invite_serializer.validated_data["email"]
+        name = invite_serializer.validated_data["body"]["name"]
+
+        parent_profile = get_parent_profile_or_create_new_user(phone=parent_phone.__str__(), email=str(email),
```
Comment (IlyaBat9) [inline open_schools_platform/organization_management/circles/views.py:135]: Why we use several method to cast to string?

File: open_schools_platform/parent_management/parents/services.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+from open_schools_platform.tasks.tasks import send_mail_to_new_user_with_celery
+
+
+def get_parent_profile_or_create_new_user(phone: str, email: str, circle_name: str) -> ParentProfile:
+    user = get_user(filters={"phone": phone})
+
+    if not user:
+        pwd = generate_user_password()
+        name = 'Родитель'
+        send_mail_to_new_user_with_celery.delay('Приглашение в кружок',
```
Comment (IlyaBat9) [inline open_schools_platform/parent_management/parents/services.py:17]: https://docs.djangoproject.com/en/4.1/topics/i18n/translation/

File: templates/new_user_circle_invite_mail_form.html [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+<body>
+<p>Приветствуем, {{ name }}!
+    <br>
+    Кружок <b>{{ circle }}</b> приглашает вас к себе на занятия. <br>Логин и пароль для входа в
+        личный кабинет на <a href="https://openschools.education/#">openschools.education:</a>
+    <br><br>
+    Ваш логин: {{ login }}<br>
+    Ваш пароль: {{ password }}<br><br>
+
+    <small>С уважением,<br> "Открытые Школы"</small>
```
Comment (IlyaBat9) [inline templates/new_user_circle_invite_mail_form.html:14]: ^^

**Review by IlyaBat9 — CHANGES_REQUESTED** [review]
There are some questions.

File: open_schools_platform/organization_management/circles/services.py [hunk 4]
```diff
@@ -44,6 +48,43 @@ def create_circle(name: str, organization: Organization, description: str, capac
...
+        family_access = user.has_perm('families.family_access', query.recipient)
+
+        if not family_access:
+            if new_status != Query.Status.CANCELED:
+                raise NotAcceptable("Circle can only set canceled status")
+        elif not circle_access:
+            if query.status != Query.Status.SENT:
+                raise NotAcceptable("Сan no longer change the query")
+            if new_status == Query.Status.CANCELED:
+                raise NotAcceptable("User cannot cancel query, he can only decline or accept it")
```
Comment (zavarin-michael) [inline open_schools_platform/organization_management/circles/services.py:73]: what if we haven't access to family and circle both?
should we raise PermissionDenied?

File: open_schools_platform/organization_management/circles/services.py [hunk 5]
```diff
@@ -44,6 +48,43 @@ def create_circle(name: str, organization: Organization, description: str, capac
...
+
+        query_update(query=query, data={"status": new_status})
+        if query.status == Query.Status.ACCEPTED:
+            if query.body is None:
+                raise MethodNotAllowed("put", detail="Query is corrupted")
+            query.body.circle = query.sender
+            query.body.family = query.recipient
```
Comment (zavarin-michael) [inline open_schools_platform/organization_management/circles/services.py:80]: query.body.family = query.recipient?
I think this should be:
query.body.student_profile = query.recipient

File: open_schools_platform/organization_management/circles/views.py [hunk 6]
```diff
@@ -108,3 +113,34 @@ def get(self, request, pk):
...
+        tags=[SwaggerTags.ORGANIZATION_MANAGEMENT_CIRCLES],
+        request_body=CircleStudentInviteSerializer,
+        responses={201: swagger_dict_response({"query": QueryStatusSerializer()})},
+        operation_description="Creates invite student query.",
+    )
+    def post(self, request, pk) -> Response:
+        invite_serializer = CircleStudentInviteSerializer(data=request.data)
+        invite_serializer.is_valid(raise_exception=True)
+
+        circle = get_circle(filters={"id": pk})
```
Comment (zavarin-michael) [inline open_schools_platform/organization_management/circles/views.py:129]: use get_circle(filters={"id": pk}, user = request.user) there

File: open_schools_platform/parent_management/parents/services.py (hunk 2 above)
Comment (zavarin-michael) [reply open_schools_platform/parent_management/parents/services.py:17]: We don't think about translation right now. It will be diffrent task in future

Comment (zavarin-michael) [reply open_schools_platform/parent_management/parents/services.py:17]: I think we should endure 'Приглашение в кружок' in special constant where we can change when we need

File: open_schools_platform/parent_management/parents/services.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        user = create_user(phone=phone, password=pwd, name=name, email=email)
+
+    return user.parent_profile
+
+
+def get_parent_family_or_create_new(parent_profile: ParentProfile):
+    family = get_family(filters={"parent_profiles": str(parent_profile.id)})
+
+    if not family:
+        family = create_family(parent=parent_profile)
```
Comment (zavarin-michael) [inline open_schools_platform/parent_management/parents/services.py:31]: it strange we leave name of family empty
it is not good, can we somehow fill it?

File: open_schools_platform/student_management/students/services.py [hunk 8]
```diff
@@ -91,6 +92,15 @@ def create_studentprofileicrcle_additional(text: str = None, parent_phone: Phone
     return additional
 
 
+def get_student_profile_or_create_new(student_phone: str, student_name: str):
+    student_profile = get_student_profile(filters={"phone": student_phone})
```
Comment (zavarin-michael) [inline open_schools_platform/student_management/students/services.py:96]: we should search student_profile by family
give a better name to method

File: open_schools_platform/student_management/students/services.py [hunk 9]
```diff
@@ -91,6 +92,15 @@ def create_studentprofileicrcle_additional(text: str = None, parent_phone: Phone
...
+
+    if not student_profile:
+        student_profile = create_student_profile(name=student_name, age=0, phone=student_phone)
```
Comment (zavarin-michael) [inline open_schools_platform/student_management/students/services.py:99]: it strange to create student_profile without name
partly this is a product error so ask product team for help

File: open_schools_platform/tasks/tasks.py [hunk 10]
```diff
@@ -6,7 +6,7 @@
 
 
 @app.task
-def send_mail_to_new_user_with_celery(subject, html_kwargs, from_email, to):
+def send_mail_to_new_user_with_celery(subject, html_kwargs, from_email, to, template='new_user_invite_mail_form.html'):
```
Comment (zavarin-michael) [inline open_schools_platform/tasks/tasks.py:9]: it is better to delete ='new_user_invite_mail_form.html'
change function in other place by insert 'new_user_invite_mail_form.html'

File: templates/new_user_circle_invite_mail_form.html (hunk 3 above)
Comment (zavarin-michael) [reply templates/new_user_circle_invite_mail_form.html:14]: ^^)

**Review by zavarin-michael — CHANGES_REQUESTED** [review]

File: open_schools_platform/organization_management/circles/services.py (hunk 4 above)
Comment (ElizavetaPetrova) [reply open_schools_platform/organization_management/circles/services.py:73]: I dont think that such a situation is possible when someone who is neither the recipient nor the sender nor the admin can somehow access the information about query and change its status.

File: open_schools_platform/parent_management/parents/services.py (hunk 7 above)
Comment (ElizavetaPetrova) [reply open_schools_platform/parent_management/parents/services.py:31]: It won't be empty. If we didn't pass family name in creation function the name will be "Family of + parent name"

File: open_schools_platform/student_management/students/services.py (hunk 8 above)
Comment (ElizavetaPetrova) [reply open_schools_platform/student_management/students/services.py:96]: (filters={"family": family.student_profiles}) like this?

File: open_schools_platform/organization_management/circles/services.py (hunk 5 above)
Comment (ElizavetaPetrova) [reply open_schools_platform/organization_management/circles/services.py:80]: student_profile don't work. Maybe query.body.student = query.recipient? Тn create_query func body_model_name = "student" or I need to change it too?

File: open_schools_platform/student_management/students/services.py (hunk 8 above)
Comment (ElizavetaPetrova) [reply open_schools_platform/student_management/students/services.py:96]: or get all student_profiles by family id?

**Review by zavarin-michael — APPROVED** [review]

**Review by zavarin-michael — APPROVED** [review]

**Review by zavarin-michael — APPROVED** [review]

**Pull request merged**